### PR TITLE
Asymmetric SSO cookies

### PR DESCRIPTION
**Pull request by Dantemss**
Leaving the old SSO code in there for now but we should remove it once we update the other apps.

Needs https://github.com/openstax/tutor-deployment/pull/529

### CODE REVIEW

File: lib/user_session_management.rb [hunk 1]
```diff
@@ -39,16 +42,17 @@ def sign_in!(user, security_log_data: {})
 
     if @current_user.is_anonymous?
       # Clear the SSO cookie
-      sso_cookies.delete(sso_cookie_name)
+      sso_cookie_jar.delete
+      old_sso_cookie_jar.delete('ox')
```
Comment (jpslav) [inline lib/user_session_management.rb]: ditto above

File: lib/user_session_management.rb [hunk 2]
```diff
@@ -39,16 +42,17 @@ def sign_in!(user, security_log_data: {})
...
 
       security_log :sign_out, security_log_data
     else
       session[:last_admin_activity] = DateTime.now.to_s if @current_user.is_administrator?
 
       # Set the SSO cookie
-      sso_cookies.permanent[sso_cookie_name] = {
-        value: { 'user' => Api::V1::UserRepresenter.new(@current_user).to_hash }
-      }
+      cookie = { value: { 'user' => Api::V1::UserRepresenter.new(@current_user).to_hash } }
```
Comment (jpslav) [inline lib/user_session_management.rb]: Possible to go ahead and add faculty status here?  Will help other efforts.  I think it is not included in the UserRepresenter by default.

File: lib/user_session_management.rb [hunk 3]
```diff
@@ -39,16 +42,17 @@ def sign_in!(user, security_log_data: {})
...
+      sso_cookie_jar.cookie = cookie
+      old_sso_cookie_jar.permanent['ox'] = cookie
```
Comment (jpslav) [inline lib/user_session_management.rb]: ditto above

File: lib/user_session_management.rb (hunk 2 above)
Comment (Dantemss) [reply lib/user_session_management.rb]: I think I have to pass in include_private_data

File: lib/user_session_management.rb [hunk 4]
```diff
@@ -1,32 +1,35 @@
...
+    cookie_jar.sso
   end
 
   # Always return an object
   def current_user
-    # The SSO cookie has sole responsbility for managing the current user
+    # The SSO cookie has sole responsibility for managing the current user
     @current_user ||= begin
-      sso_cookie = sso_cookies[sso_cookie_name]
+      sso_cookie = sso_cookie_jar.cookie || old_sso_cookie_jar['ox']
```
[the remark this answers is not in the document]
Comment (Dantemss) [reply lib/user_session_management.rb]: OK I'll leave that in the secrets

File: lib/user_session_management.rb (hunk 2 above)
Comment (jpslav) [reply lib/user_session_management.rb]: Kind of don’t want other data in there eg all emails. I can look at this later. Would be cool if we could just have fac status. Would let other apps enable things without callback to accounts. Maybe faculty status isn’t really private?

Comment (jpslav) [reply lib/user_session_management.rb]: Kind of don’t want other data in there eg all emails. I can look at this later. Would be cool if we could just have fac status. Would let other apps enable things without callback to accounts. Maybe faculty status isn’t really private?

Comment (Dantemss) [reply lib/user_session_management.rb]: Yeah maybe faculty status is not private (being faculty somewhere is pretty much public knowledge anyway).

**Comment by Dantemss** [comment]
The plan is to sign the cookie with the signature private key (readable with the signature public key) then encrypt it with the encryption public key (readable with the encryption private key).

Or we could just do asymmetric signature with symmetric encryption, since the public key can be generated from the private key anyway...

**Comment by Dantemss** [comment]
Sure, but Travis exploded. I think it's just intermittent stuff, so I'm rerunning

**Pull request merged**